src/keychain_mac.cpp

```diff
@@ -101,67 +102,80 @@ void updateError(keychain::Error &err, OSStatus status) {
     }
 }
 
-void handleCFCreateFailure(keychain::Error &err,
-                           const std::string &errorMessage) {
+void setGenericError(keychain::Error &err, const std::string &errorMessage) {
+    err = keychain::Error{};
     err.message = errorMessage;
     err.type = keychain::ErrorType::GenericError;
     err.code = -1;
 }
 
-CFStringRef createCFStringWithCString(const std::string &str,
-                                      keychain::Error &err) {
-    CFStringRef result = CFStringCreateWithCString(
-        kCFAllocatorDefault, str.c_str(), kCFStringEncodingUTF8);
-    if (result == NULL) {
-        handleCFCreateFailure(err, "Failed to create CFString");
+/*! \brief Helper to manage the lifetime of CFObjects */
+template <typename T> struct ScopedCFRef {
+  public:
+    explicit ScopedCFRef(T &&ref) : _ref(std::move(ref)) {}
+    ~ScopedCFRef() {
+        if (_ref)
+            CFRelease(_ref);
     }
+
+    ScopedCFRef(ScopedCFRef &&) noexcept = default;
```

Comment:
```suggestion
    ScopedCFRef(ScopedCFRef &&other) noexcept : _ref(other._ref) { other._ref = nullptr; }
```

... after moving, the moved-from object is most probably still being destructed. You want to make sure that the moved-from object doesn't call `CFRelease()` causing a double-free.

From you CMakeLists.txt I believe you're on C++11. With C++14 you could formulate this somewhat nicer:

```C++
ScopedCFRef(ScopedCFRef &&other) noexcept : _ref(std::exchange(other._ref, nullptr)) {}
```